### DISCUSSION TITLE
Emoji on Windows vs macOS

### DISCUSSION THREAD

**Discussion by waveofthought-code** [airsdk/Adobe-Runtime-Support#657] [Q&A]
I'm using _sans as a device font (unembedded) in a standard flash.text.TextField

**On Windows builds, I get simple black emojis.
On macOS, I get the complex colored emoji.**

I would accept this as a normal thing if it wasn't for the following:

_Pre-2021, when my app was run in any Windows browser, the rendered emoji was in color, using the official Windows colored glyphs!_

I can only imagine that browser manufacturers were doing something under the hood in order to translate _sans emoji glyphs in to their colored counterparts, so i'm hoping there's something Harman could do here with the Air SDK as well.

Maybe there's a trick i'm unaware of?   Embedding an "emoji font" simply doesnt work.. I've been down that road many times before.   I'd like to continue using _sans as my catchall for OS-supported emoji if at all possible!

For what its worth I've tried Air32 and all versions of Air33 up to .345

Thanks!

**Comment by Mintonist** [comment]
I have a chat in my game (platforms: Android, IOS, Windows). What I need to integrate default smiles (from software device clipboard) that can be shown in all platforms?
It is enough to use _sans font for feather TextInput (I use StageTextRender)?
Pls, tell about yours experience. Thx!

**Reply by waveofthought-code** [reply]
I don't use the StageTextRenderer from Feathers so I can't comment on that, but from my experience using _sans in a flash.text.TextField, i get *ALL* the ios-style colored emoji on macOS, but all the B+W simple emoji on Windows (unfortunately, which is why i made this post)

copy/paste (clipboard) works fine on both platforms.

**Comment by JohnBlackburne** [comment]
What sort of app is it? From what you write it's running in a browser so a Flash plugin app?

Flash plugin was retired at the end of 2020. The exact date varied; some browsers disabled it earlier (Safari did it months ago), some waited until January 12. But by now it should not be working anywhere. I wonder if it's something Windows is doing, disabling support for certain OS features for Flash.

Other than that I can't think what changed, and you wrote that it's not the AIR version, or a particular browser. The date is particularly suggestive of something to do with the Flash EOL. I would also try it on other Windows devices, in case it's a glitch with a particular PC.

**Reply by waveofthought-code** [reply]
My bad i thought it was self-explanaotry that i'm now running in AIR, since this is the Harman AIR SDK support forum.

To clarify, when my app *was* in flashplayer, the emoji worked great, presumably because the browsers are facilitating it somehow.
Now that i'm in Air, i don't seem to have that same luxury.

I'm looking for a solution for Air.

**Reply by JohnBlackburne** [reply]
Ah that makes more sense. I don't have any experience of desktop AIR apps. On mobile it's a completely different platform; the classic API in particular is much less suited for mobile use, you need to use Stage3D for drawing. I don't know what's different on desktop – clearly something is !

**Comment by Mintonist** [comment]
May it be helpfully for somebody.
Here is my code. I use StageTextTextEditor instead of existing simple Starling TextField (_mcItem.t_msg) to correctly display smiles|emoji. It works across Android/IOS/Windows platforms. For Windows I use editor.fontFamily = "Segoe UI Symbol". This system fonts display black&white emoji.

```
var txtInput:TextInput = new TextInput();
			txtInput.x = _mcItem.t_msg.x;
			txtInput.y = _mcItem.t_msg.y;
			txtInput.width = _mcItem.t_msg.width;
			txtInput.height = _mcItem.t_msg.height;
			txtInput.text = _mcItem.t_msg.text;
			txtInput.verticalAlign = "top";
			txtInput.textEditorFactory = function ():ITextEditor {
				var editor:StageTextTextEditor = new StageTextTextEditor();
				editor.fontFamily = null; // to use device font on mobile. Use "Segoe UI Symbol" on Windows
				editor.fontSize = 20;
				editor.fontWeight = FontWeight.BOLD;
				editor.color = color;
				editor.isEditable = false;
				editor.multiline = true;
				editor.textAlign = TextFormatAlign.LEFT;
				return editor;
			};
			_mcItem.removeChild(_mcItem.t_msg);
			_mcItem.addChild(txtInput);
			_mcItem.t_msg = txtInput;
			_mcItem.t_msg.touchable = false;
```
